src/Nexus/Services/PipelineService.cs

```diff
@@ -0,0 +1,146 @@
+        {
+            var id = Guid.NewGuid();
+
+            pipelineMap.AddOrUpdate(
+                id,
+                pipeline,
+                (key, _) => pipeline
+            );
+
+            return id;
+        }, saveChanges: true);
+    }
+
+    public bool TryGet(
+        string userId,
+        Guid pipelineId,
+        [NotNullWhen(true)] out DataSourcePipeline? pipeline)
+    {
+        var pipelineMap = GetPipelineMap(userId);
+
+        return pipelineMap.TryGetValue(pipelineId, out pipeline);
+    }
+
+    public Task DeleteAsync(string userId, Guid pipelineId)
+    {
+        return InteractWithPipelineMapAsync<object?>(userId, pipelineMap =>
+        {
+            var pipelineEntry = pipelineMap
+                .FirstOrDefault(entry => entry.Key == pipelineId);
+
+            pipelineMap.TryRemove(pipelineEntry.Key, out _);
+            return default;
+        }, saveChanges: true);
+    }
+
+    public async Task<IReadOnlyDictionary<string, IReadOnlyDictionary<Guid, DataSourcePipeline>>> GetAllAsync()
+    {
+        var result = new Dictionary<string, IReadOnlyDictionary<Guid, DataSourcePipeline>>();
+
+        foreach (var userId in _databaseService.EnumerateUsers())
+        {
+            var pipelines = await GetAllForUserAsync(userId);
+            result[userId] = pipelines;
+        }
+
+        return result;
+    }
+
+    public Task<IReadOnlyDictionary<Guid, DataSourcePipeline>> GetAllForUserAsync(
+        string userId)
+    {
+        return InteractWithPipelineMapAsync(
+            userId,
+            pipelineMap => (IReadOnlyDictionary<Guid, DataSourcePipeline>)pipelineMap,
+            saveChanges: false);
+    }
+
+    private ConcurrentDictionary<Guid, DataSourcePipeline> GetPipelineMap(
+        string userId)
+    {
+        return _cache.GetOrAdd(
+            userId,
+            key =>
+            {
+                if (_databaseService.TryReadPipelineMap(userId, out var jsonString))
+                {
+                    return JsonSerializer.Deserialize<ConcurrentDictionary<Guid, DataSourcePipeline>>(jsonString)
+                        ?? throw new Exception("pipelineMap is null");
+                }
+
+                else
+                {
+                    return new ConcurrentDictionary<Guid, DataSourcePipeline>();
+                }
+            });
+    }
+
+    private async Task<T> InteractWithPipelineMapAsync<T>(
+        string userId,
+        Func<ConcurrentDictionary<Guid, DataSourcePipeline>, T> func,
+        bool saveChanges)
+    {
+        await _semaphoreSlim.WaitAsync().ConfigureAwait(false);
+
+        try
+        {
+            var pipelineMap = GetPipelineMap(userId);
+            var result = func(pipelineMap);
+
+            if (saveChanges)
+            {
+                using var stream = _databaseService.WritePipelineMap(userId);
+                JsonSerializerHelper.SerializeIndented(stream, pipelineMap);
```

Comment:
Because `JsonSerializer.SerializeAsync` operates on the `Stream` it gets as a parameter and that stream must operate truly `async` to have an effect. We are providing the variable named `stream` to the method and this variable is a `FileStream` which is returned from the `DatabaseService.WritePipelineMap()` method. A file stream is only async when it opened asynchronously explicitly (see the [useAsync](https://learn.microsoft.com/en-us/dotnet/api/system.io.filestream.-ctor?view=net-8.0&redirectedfrom=MSDN#system-io-filestream-ctor(system-string-system-io-filemode-system-io-fileaccess-system-io-fileshare-system-int32-system-boolean)) parameter. This is a bit overkill for a small file like `pipelines.json`. So, yes, in theory we could use `SerializeIndentedAsync` but since this requires use to keep track of async capabilities of the underyling stream I would only to this for larger files, i.e. when the performance impact becomes relevant.

Reply:
Thank you for the explanation. :-)